### DISCUSSION TITLE
Usage of  EntityTrait as Generic with full features

### DISCUSSION THREAD

**Discussion by amorfc** [SeaQL/sea-orm#1789] [Q&A]
Hello everyone, 

Currently I am dealing with my backend grpc server, trying to implement data layer and repo abstraction.

While I am using sea-orm entity types for my CRUD's and cursor pagination I am trying to implement something like base repository struct with trait features. Like don't want to write CRUD's and pagination methods for each entities. 

I started to use Generic trait implementation for my base repository but I need some components in methods. 

Here is the traits signature.

Be aware that  I am a newbie for Rust <3

Here is signature that I am trying to implement but here T does not have smth like ActiveModel. So I might use use all sea-orm entities features in here at the definition maybe like Columns,Model or others. 

Here T::Model is fine but the others giving errors.

I tought EntityTrait type will be fine because that is the trait which is provides CRUD api. But while I use insert and the others I have to use ActiveModel but the EntityTrait does not have that trait.
 

```rust
use sea_orm::{ActiveModelTrait, DbErr, EntityTrait};

pub trait Repository<T: EntityTrait> {
    fn create<U: ActiveModelTrait>(&self, entity: U) -> Result<T::Model, DbErr>;
    fn update(&self, entity: T::ActiveModel) -> Result<T::Model, DbErr>;
    fn find(&self, id: T::PrimaryKey) -> Result<Option<T::Model>, DbErr>;
}
```

Is there a better way to achieve this solution or my path is wrong ?

Thank you for all response from who enthusiastic person for Rust

**Comment by tyt2y3** [comment]
This looks pretty good.

**Reply by amorfc** [reply]
The compiler feels angry about T::ActiveModel because T did not implement ActiveModelTrait. 
EntityTrait does not implements it so ?

But I agree with you it looks so simple and pretty good I wish it is works 😌

**Comment by amorfc** [comment]
Hello again everyone ,
This is how I fix  a BaseRepositoryImpl implementation for my backend server. 
My BaseRepoTrait has to have common sea-orm entity methods like CRUD and all repositories has to implements this BaseRepoTrait to escape duplicate coding. 

Beside that all domain repositories should have extendable BaseRepoTrait to define domain repository specific method over BaseRepoTrait CRUDs. 

Here is my brief explanation of my BaseRepositoryImpl trait 

```rust
use sea_orm::{
    ActiveModelBehavior, ActiveModelTrait, DbErr, EntityTrait, IntoActiveModel, TransactionTrait,
};

use crate::infra::db_initializor::{LetDbConnection, LetDbTransaction};

pub trait RepoDbConnectionProvider {
    fn db_connection(&self) -> &LetDbConnection;
}

// Define a new trait that provides the db_transaction method
#[tonic::async_trait]
pub trait RepoDbTransactionProvider: RepoDbConnectionProvider {
    async fn db_tx(&self) -> Result<LetDbTransaction, DbErr> {
        self.db_connection().begin().await
    }
}

#[tonic::async_trait]
pub trait BaseRepositoryImpl<A, E>: RepoDbConnectionProvider + RepoDbTransactionProvider
where
    A: ActiveModelTrait + ActiveModelBehavior + Send + 'static,
    <A::Entity as EntityTrait>::Model: IntoActiveModel<A>,
    E: EntityTrait + 'static,
{
    async fn save(&self, model: A) -> Result<A, String> {
        let db_tx = self.db_connection();

        match model.save(db_tx).await {
            Ok(model) => Ok(model),
            Err(_) => return Err("Failed to save model".to_string()),
        }
    }

    async fn insert(&self, model: A) -> Result<A, String> {
        let db_tx = self.db_connection();

        match model.insert(db_tx).await {
            Ok(model) => Ok(model.into_active_model()),
            Err(_) => return Err("Failed to insert model".to_string()),
        }
    }
}

```

I used transaction pattern for my create methods because in the future if I need to perform multiple record save operation at the same time as a atomic I can call all methods without commit to the database if I had any problem in the any operation.

Here is my concrete UserRepository type for my BaseRepositoryImpl and needed super traits.

```rust
use entity::user as UserEntity;

use crate::{
    application::repositories::common::repository::{
        BaseRepositoryImpl, RepoDbConnectionProvider, RepoDbTransactionProvider,
    },
    infra::db_initializor::LetDbConnection,
};

// Define a new trait that provides the db_connection method

pub struct UserRepository {
    db_connection: LetDbConnection,
}

#[tonic::async_trait]
pub trait UserRepositoryTrait
where
    Self: BaseRepositoryImpl<UserEntity::ActiveModel, UserEntity::Entity>,
{
    fn new(db_connection: LetDbConnection) -> Self;
    async fn create_user(&self, user: UserEntity::ActiveModel) -> Result<(), String>;
}

#[tonic::async_trait]
impl UserRepositoryTrait for UserRepository {
    fn new(db_connection: LetDbConnection) -> Self {
        Self { db_connection }
    }

    async fn create_user(&self, model: UserEntity::ActiveModel) -> Result<(), String> {
        self.create(model).await?;

        Ok(())
    }
}

// Then, when you implement the trait, you can specify the types:
#[tonic::async_trait]
impl BaseRepositoryImpl<UserEntity::ActiveModel, UserEntity::Entity> for UserRepository {}

impl RepoDbConnectionProvider for UserRepository {
    fn db_connection(&self) -> &LetDbConnection {
        &self.db_connection
    }
}

impl RepoDbTransactionProvider for UserRepository {}
```

If anyone need a implementation  smt like this can refer from here.

But if you see any issue in this implementation please let me know thanks :)
